crates/qpy/src/circuits.rs

```diff
@@ -0,0 +1,2166 @@
+use qiskit_circuit::imports;
+use qiskit_circuit::operations::{ArrayType, Operation, OperationRef, Param, StandardInstruction};
+use qiskit_circuit::packed_instruction::{PackedInstruction, PackedOperation};
+use qiskit_circuit::{Clbit, Qubit};
+
+use smallvec::SmallVec;
+use uuid::Uuid;
+
+use crate::annotations::AnnotationHandler;
+use crate::bytes::Bytes;
+use crate::consts::standard_gate_from_gate_class_name;
+use crate::formats;
+use crate::params::{pack_param, pack_param_obj, unpack_param};
+use crate::value::{
+    circuit_instruction_types, deserialize, deserialize_with_args, dumps_value,
+    expression_var_declaration, get_circuit_type_key, pack_generic_data, pack_standalone_var,
+    serialize, tags, unpack_generic_data, DumpedValue, ExpressionType, QPYReadData, QPYWriteData,
+};
+
+use crate::UnsupportedFeatureForVersion;
+
+const UNITARY_GATE_CLASS_NAME: &str = "UnitaryGate";
+// used when writing
+//used when reading
+#[derive(Debug)]
+struct CustomCircuitInstructionData {
+    gate_type: u8,
+    num_qubits: u32,
+    num_clbits: u32,
+    definition_circuit: Option<Py<PyAny>>,
+    // num_ctrl_qubits: u32,
+    // ctrl_state: u32,
+    base_gate_raw: Bytes,
+}
+
+pub mod register_types {
+    pub const QREG: u8 = b'q';
+    pub const CREG: u8 = b'c';
+}
+
+fn getattr_or_none<'py>(
+    py_object: &'py Bound<'py, PyAny>,
+    name: &str,
+) -> PyResult<Option<Bound<'py, PyAny>>> {
+    if py_object.hasattr(name)? {
+        let attr = py_object.getattr(name)?;
+        if attr.is_none() {
+            Ok(None)
+        } else {
+            Ok(Some(attr))
+        }
+    } else {
+        Ok(None)
+    }
+}
+fn is_python_gate(py: Python, op: &PackedOperation, python_gate: &Bound<PyAny>) -> PyResult<bool> {
+    match op.view() {
+        OperationRef::Gate(pygate) => {
+            if pygate.gate.bind(py).is_instance(python_gate)? {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+        _ => Ok(false),
+    }
+}
+
+pub mod bit_types {
+    pub const QUBIT: u8 = b'q';
+    pub const CLBIT: u8 = b'c';
+}
+
+fn get_packed_bit_list(
+    inst: &PackedInstruction,
+    circuit_data: &CircuitData,
+) -> Vec<formats::CircuitInstructionArgPack> {
+    let mut result: Vec<formats::CircuitInstructionArgPack> = Vec::new();
+    for qubit in circuit_data.get_qargs(inst.qubits).iter() {
+        result.push(formats::CircuitInstructionArgPack {
+            bit_type: bit_types::QUBIT,
+            index: (qubit.index() as u32),
+        });
+    }
+    for clbit in circuit_data.get_cargs(inst.clbits).iter() {
+        result.push(formats::CircuitInstructionArgPack {
+            bit_type: bit_types::CLBIT,
+            index: (clbit.index() as u32),
+        });
+    }
+    result
+}
+
+fn gate_class_name(py: Python, op: &PackedOperation) -> PyResult<String> {
+    let name = match op.view() {
+        // getting __name__ for standard gates and instructions should
```

Comment:
Agreed, we should have a Python-free mapping whenever possible. I opened this: #15095